### ISSUE TITLE
Multiline associative operator usage looks strange

### ISSUE THREAD

**Issue by tysonzero** [fourmolu/fourmolu#168]
For example parsers:

```
parser = Foo <$> foo
     <|> Bar <$> baz <*> baz
     <|> Qux <$> tux

parser =
    Foo <$> foo
        <|> Bar <$> bar <*> baz
        <|> Qux <$> qux
```

Also chained lens updates

```
update = foo .~ 5
     >>> bar ?~ "new"
     >>> baz %~ (+ 1)

update =
    foo .~ 5
        >>> bar ?~ "new"
        >>> baz %~ (+ 1)
```

An option to be able keep the former would be fantastic.

**Comment by brandonchinn178** [comment]
Can this be pushed upstream to Ormolu? Fourmolu doesn't do anything different here. If Ormolu gets a fix for this, we'll get it automatically when we merge Ormolu changes back in

**Comment by tysonzero** [comment]
Is the plan to indefinitely maintain codebase compatibility with Ormolu? Won't that become increasingly difficult the more configurable Fourmolu is?

Building a less configurable formatter on top of a more configurable formatter is of course easy, but the other way around seems challenging.

Regardless yes it seems like this can be upstreamed. Have you double checked that the output of Ormolu is the same? I don't have it installed.

**Comment by brandonchinn178** [comment]
> Is the plan to indefinitely maintain codebase compatibility with Ormolu?

Yes. Technically, it wouldn't be "codebase compatibility" because it's not like Ormolu can use Fourmolu code as a drop-in replacement, but yes, Fourmolu will always be pulling upstream changes + merging them into our more configurable codebase.

> Won't that become increasingly difficult the more configurable Fourmolu is?

Sure, but this way, Fourmolu _only_ has to worry about configuration-related changes. If we pretended Fourmolu never added any more configuration options, the _only_ development work in Fourmolu is to pull in Ormolu changes and make it work with our options. And whatever formatting improvements or changes Ormolu makes, we probably want to do the same thing, so it would be just as much work if we never merged in Ormolu changes and manually reimplement all of their PRs into Fourmolu.

> Have you double checked that the output of Ormolu is the same? I don't have it installed.

Yes, this is ormolu's behavior:
```hs
parser =
  Foo <$> foo
    <|> Bar <$> baz <*> baz
    <|> Qux <$> tux

update =
  foo .~ 5
    >>> bar ?~ "new"
    >>> baz %~ (+ 1)
```

I'm closing this ticket, because this is an Ormolu issue

**Comment by tysonzero** [comment]
Looks like the ormolu folks are taking the rather strange stance that the latter way of formatting things is actually reasonable, so if this is to be supported it may have to be fourmolu-only code.

**Comment by brandonchinn178** [comment]
I'm leaning towards Ormolu's stance here too. The only reason things are lined up nicely is because of the indentation size and the operators being three characters long. With indentation=2, it looks less nice:

```hs
my_parser = Foo <$> foo
  <|> Bar <$> bar
  <|> Baz <$> baz
```

My counterpoint to your "why is the first field single'd out" is "why should operators be any different from other expressions?" Which do you think looks better:
```hs
my_abs x = if x > 0
  then -x
  else x

my_abs x =
  if x > 0
    then -x
    else x
```

**Comment by tysonzero** [comment]
For `my_parser` I'd do:

```
my_parser = Foo <$> foo
        <|> Bar <$> bar
        <|> Baz <$> baz
```

Both of the `my_abs` options look reasonable enough to me, can't say I have much preference between the two.

Isn't the idea behind `fourmolu` to be configurable? Even if this isn't the style you would use, is there any reason it should be explicitly not allowed as an option? Are there other formatters that go further in the direction of configurability than fourmolu?

**Comment by brandonchinn178** [comment]
I'm a bit surprised you're fine with both `my_abs` formats 😄 To me, the first looks absolutely egregious. But let's put that aside; your comment clarified what you're asking for.

I see now that you're not merely asking about the hanging behavior (i.e. whether the first line in the expression should stay on the initial line and the rest of the lines should hang below). You're asking for the expressions to vertically align. Another example of this issue is:

```hs
-- input
foo x =
    doThing
  . otherThing
  . another
  $ x

-- reformatted
foo x =
  doThing
    . otherThing
    . another
    $ x
```

So to be clear, do you just want the vertical alignment, or do you also want to keep the first line on the previous line? Doing just the vertical alignment would be doable; I see that as much less work than doing both (hanging expressions is a much more systemic issue than this). I would be open to a PR that just realigns the expressions vertically, while keeping the line break logic the same.

---

Some meta points:
* FWIW, this wasn't clear in the original issue. In hindsight, it would've been clearer to mention exactly what you wanted (vertical alignment) instead of just a vague "it looks strange" (too subjective; it didn't look strange to me or Ormolu at first!)
* "Isn't the idea behind fourmolu to be configurable?" - Sure, but that doesn't mean we accept absolutely every configuration under the sun. Each configuration option adds complexity to the formatting logic (which can cause unintuitive behavior when interacting with multiple features) and maintenance burden, by increasing the risk of conflicts when merging Ormolu
* "Are there other formatters that go further in the direction of configurability than fourmolu?" - Not that I'm aware of, but there's nothing stopping you from forking and making your own formatter with the exact style you like 😄

**Comment by tysonzero** [comment]
I see the first `my_abs` as the same as how I would format the equivalent case statement:

```
my_abs x = if x > 0
  then -x
  else x

my_abs x = case x > 0 of
  True -> -x
  False -> x
```

So I guess that's why I didn't blink at it.

---

Hmm I was really hoping for both. How would my original `parser` example work if the first expression has to go on a new line?

```
parser = Foo <$> foo
     <|> Bar <$> baz <*> baz
     <|> Qux <$> tux

parser =
        Foo <$> foo
    <|> Bar <$> baz <*> baz
    <|> Qux <$> tux
```

Some other code examples:

```
o = SetState $ s & _foo .~ foo
                 & _bar .~ bar
                     
result <- run $ mkArgs reqArg
    & _optArg1 ?~ optArg1
    & _optArg2 ?~ optArg2
```

---

Meta-commentary is fair yeah. I actually did make a formatter in the style I like (https://github.com/polimorphic/poli-format) but it's far from feature complete and trying to work with a well established one instead for new projects seemed like a better approach.

**Comment by brandonchinn178** [comment]
This is how Fourmolu today reformats that code:
```hs
parser =
  Foo <$> foo
    <|> Bar <$> baz <*> baz
    <|> Qux <$> tux

o =
  SetState $
    s
      & _foo .~ foo
      & _bar .~ bar

main = do
  result <-
    run $
      mkArgs reqArg
        & _optArg1 ?~ optArg1
        & _optArg2 ?~ optArg2
```
This is what the new rule would do, keeping the logic to put thing on the next line, but just vertically re-align the operators:
```hs
parser =
      Foo <$> foo
  <|> Bar <$> baz <*> baz
  <|> Qux <$> tux

o =
  SetState $
      s
    & _foo .~ foo
    & _bar .~ bar

main = do
  result <-
    run $
        mkArgs reqArg
      & _optArg1 ?~ optArg1
      & _optArg2 ?~ optArg2
```
